### PR TITLE
Hidden Power

### PR DESCRIPTION
**Pull request by VeteranPadgett**
Here is what I have managed to do so far for Hidden Power. I still need to check if the type is sent to the battle server correctly and make the teambuilder load hidden power type of the saved pokes.

### CODE REVIEW

**Comment by VeteranPadgett** [comment]
Pokemon doesn't have right hidden power type in battle :(

**Comment by VeteranPadgett** [comment]
With this commit everything client side is working now. Just need to send the hiddenpowertype stored on client to the server.

File: src/com/podevs/android/poAndroid/poke/TeamPoke.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
 	public short pokeball;
 	public short ability;
 	public byte nature;
+	public byte hiddenPowerType = 16;
```
Comment (turbedi) [inline src/com/podevs/android/poAndroid/poke/TeamPoke.java]: Is there a reason that this variable is declared, but the others aren't?

File: src/com/podevs/android/poAndroid/poke/TeamPoke.java [hunk 2]
```diff
@@ -265,6 +272,10 @@ public void serializeBytes(Baos bytes) {
 			if (gen.num > 2 && happiness != 0) {
 				b.write(happiness);
 			}
+
+			if (gen.num > 6 && hiddenPowerType() != 16) {
```
Comment (turbedi) [inline src/com/podevs/android/poAndroid/poke/TeamPoke.java]: Is the hiddenPowerType check really necessary?

**Comment by turbedi** [comment]
Does the PR contain changes that attempt to send the Hidden Power type to the server and it isn't working, or are the changes only on the client side?

File: src/com/podevs/android/poAndroid/teambuilder/MoveChooserFragment.java [hunk 3]
```diff
@@ -103,25 +103,30 @@ public boolean onItemLongClick(AdapterView<?> parent, View view, int position, l
...
+					else if (poke().validHiddenPowerType(type)) {
+						poke().hiddenPowerType = (byte)type;
+					} else {
+						type = poke().hiddenPowerType();
+					}
+
+					if (listener != null) {
+						listener.onMovesetChanged(true);
+					}
+					moveButton.reloadImage(type, view);
```
Comment (MidwayMarshall) [inline src/com/podevs/android/poAndroid/teambuilder/MoveChooserFragment.java]: there is already a function that will achieve this: moveAdapter.notifyDataSetChanged();

**Comment by VeteranPadgett** [comment]
I declared the hidden power to make it dark type by default but it may be unnecessary. The hidden power type check is to match what I did on the server. We can probably get rid of it as long as that gets changed for the server as well.

I attempted to send the type to the server. I believe it uses the team poke serialize. So far the type always ends up being based on ivs though like in previous gens. This means either hidden power dark is being sent or no type sent at all.

**Comment by MidwayMarshall** [comment]
Is the desktop/server communication working correctly in regards to sending hidden power?

**Comment by VeteranPadgett** [comment]
Ok, I didn't know about the built in function. Thanks! I'm busy right now, but I'll change it when I get home this evening.

**Comment by VeteranPadgett** [comment]
Yes, pc to server works fine.

**Comment by VeteranPadgett** [comment]
Updated buildHiddenPowerDialog()

Now that I think about the check for hiddenPower != 16 it matters for the server to not send hiddenpower to older clients, and we have to do it on android to match what data the server will be sending in.

**Comment by VeteranPadgett** [comment]
Finally works! Was missing setting the hasHiddenPower flag to send to server.

Make sure to get the latest server source to test this. I moved where the hiddenPower flag is in the main source enum, since it was in a confusing spot that didn't match the order the flags were used.

**Comment by turbedi** [comment]
Can you remove the reloadImage function if it is now unused?
Besides that, I am fine with merging it if you have tested it.

**Comment by coyotte508** [comment]
The hidden power flag was added last, it should be last, otherwise it breaks when connecting to old servers. Well android client might break ayway if protocol version isn't properly checked but computer clients don't. The old flags keep their order and the new one is added after that.

File: src/com/podevs/android/poAndroid/poke/TeamPoke.java (hunk 1 above)
Comment (coyotte508) [inline src/com/podevs/android/poAndroid/poke/TeamPoke.java]: Is there no enum (Like Type.Dark)?

File: src/com/podevs/android/poAndroid/battle/BattlePoke.java [hunk 4]
```diff
@@ -21,7 +21,7 @@
 	// byte statusCount = 0;
 	// byte originalStatusCount = 0;
 	byte nature = 0;
-	byte hiddenPower = 0;
+	byte hiddenPower = 16;
```
Comment (coyotte508) [inline src/com/podevs/android/poAndroid/battle/BattlePoke.java]: Should not use a direct numeral. Should use `TypeInfo.Type.Dark.ordinal()`.

File: src/com/podevs/android/poAndroid/poke/TeamPoke.java (hunk 1 above)
Comment (coyotte508) [inline src/com/podevs/android/poAndroid/poke/TeamPoke.java]: And it should be `TypeInfo.Type.Dark.ordinal()`.

File: src/com/podevs/android/poAndroid/poke/TeamPoke.java [hunk 5]
```diff
@@ -238,8 +245,8 @@ public void setGen(Gen gen) {
 
 	public void serializeBytes(Baos bytes) {
 		Baos b = new Baos();
-//								hasGen, hasNickname, 	hasPokeball, hasHappiness, hasPPups,  hasIVs,
-		b.putFlags(new boolean[]{true, nick.length() > 0, false, 	happiness != 0, false, 	true});
+//								hasGen, hasNickname, hasPokeball, hasHappiness, hasHiddenPower, 	hasPPups, hasIVs,
+		b.putFlags(new boolean[]{true, nick.length() > 0, false, happiness != 0,hiddenPowerType != 16, false, true});
```
Comment (coyotte508) [inline src/com/podevs/android/poAndroid/poke/TeamPoke.java]: Flags should not be reordered, or it  breaks compatibility

File: src/com/podevs/android/poAndroid/poke/TeamPoke.java [hunk 6]
```diff
@@ -124,6 +129,7 @@ public void reset() {
 		item = 0;
 		ability = 0;
 		nature = 0;
+		hiddenPowerType = 16;
```
Comment (coyotte508) [inline src/com/podevs/android/poAndroid/poke/TeamPoke.java]: Use the enum Dark...

File: src/com/podevs/android/poAndroid/poke/TeamPoke.java [hunk 7]
```diff
@@ -88,9 +89,13 @@ public void loadFromBais(Bais msg) {
 			if (gen.num > 2 && network.readBool()) { //happiness flag
 				happiness = b.readByte();
 			}
+			if (gen.num > 6 && network.readBool()) { //hidden power flag
```
Comment (coyotte508) [inline src/com/podevs/android/poAndroid/poke/TeamPoke.java]: Either move that below or read all the flag at the start.

File: src/com/podevs/android/poAndroid/poke/TeamPoke.java [hunk 8]
```diff
@@ -177,6 +183,7 @@ public void setNum(UniqueID id) {
 
 			if (fullReset) {
 				happiness = 0;
+				hiddenPowerType = 16;
```
Comment (coyotte508) [inline src/com/podevs/android/poAndroid/poke/TeamPoke.java]: Use the enum

**Comment by coyotte508** [comment]
Also the [protocol version](https://github.com/po-devs/android-client/blob/master/src/com/podevs/android/poAndroid/ProtocolVersion.java) should be updated so the server knows it is a new client. Otherwise it will behave with it as if it was an old client and some gen 7 stuff might not go through. The current version is 4 (major) / 0 (minor)

**Comment by VeteranPadgett** [comment]
Took care of flags. Updating the protocol would be a lot of work just so that the battle move button on android would be the same color as hidden power type. I don't think android does this even with the past gens, so protocol changes alone wouldn't do anything.

I'll work on it, but I wouldn't wait for it on this pr. I'm going to be very busy next week, so it probably wouldn't get done until January.

**Comment by coyotte508** [comment]
Yea, I checked and protocol change would only allow client to receive hiddenpower and nature in battle. Looks like the rest is still compatible!

**Pull request merged**